File: mockscull/src/numscull/crypto.py
"""Encryption layer: NaCl Box key exchange and encrypted channel."""

import json
import os
import socket
import struct
from pathlib import Path
from typing import Any, Dict

from nacl.bindings import crypto_box, crypto_box_open, crypto_scalarmult_base

from .transport import (
    BLOCK_SIZE,
    ENCRYPTED_BLOCK_SIZE,
    HEADER_SIZE,
    KEY_LEN,
    NONCE_LEN,
    pack_plaintext_bytes,
    read_exact,
)


def counter_nonce(counter: int) -> bytes:
    """Encode counter as 24-byte nonce: LE u64 + 16 zero bytes."""
    return struct.pack("<Q", counter) + b"\x00" * 16


def generate_x25519_keypair() -> tuple[bytes, bytes]:
    """Generate an X25519 keypair. Returns (public_key, secret_key)."""
    sk = os.urandom(32)
    pk = crypto_scalarmult_base(sk)
    return pk, sk


def load_keypair(identity_name: str, config_dir: Path) -> tuple[bytes, bytes]:
    """Returns (public_key, secret_key), each 32 bytes."""
    identity_path = config_dir / "identities" / identity_name
    raw = identity_path.read_bytes()
    if len(raw) != 64:
        raise ValueError(f"Expected 64-byte identity file, got {len(raw)}")
    return raw[:32], raw[32:]


class EncryptedChannel:
    """Encrypted communication using ephemeral X25519 + NaCl Box."""

    def __init__(
        self,
        sock: socket.socket,
        ours_recv_sk: bytes,
        ours_send_sk: bytes,
        theirs_recv_pk: bytes,
        theirs_send_pk: bytes,
    ):
        self.sock = sock
        self.ours_recv_sk = ours_recv_sk
        self.ours_send_sk = ours_send_sk
        self.theirs_recv_pk = theirs_recv_pk
        self.theirs_send_pk = theirs_send_pk
        self.send_nonce: int = 1
        self.recv_nonce: int = 1

    def send(self, message: Dict[str, Any]) -> None:
        json_bytes = json.dumps(message).encode("utf-8")
        framed = pack_plaintext_bytes(json_bytes)
        if len(framed) > BLOCK_SIZE - 2:
            raise ValueError(f"Message too large: {len(framed)} > {BLOCK_SIZE - 2}")

        block = bytearray(BLOCK_SIZE)
        struct.pack_into("<H", block, 0, len(framed))
        block[2 : 2 + len(framed)] = framed
        padding_start = 2 + len(framed)
        block[padding_start:] = os.urandom(BLOCK_SIZE - padding_start)

        nonce = counter_nonce(self.send_nonce)
        self.send_nonce += 1

        ct = crypto_box(bytes(block), nonce, self.theirs_send_pk, self.ours_send_sk)
        self.sock.sendall(ct)

    def recv_raw(self) -> bytes:
        """Receive and decrypt one block, returning raw payload bytes."""
        ct = read_exact(self.sock, ENCRYPTED_BLOCK_SIZE)
        nonce = counter_nonce(self.recv_nonce)
        self.recv_nonce += 1

        block = crypto_box_open(ct, nonce, self.theirs_recv_pk, self.ours_recv_sk)
        msg_len = struct.unpack("<H", block[:2])[0]
        return block[2 : 2 + msg_len]

    def recv(self) -> Dict[str, Any]:
        """Receive, decrypt, and parse a JSON response."""
        data = self.recv_raw()
        header = data[:HEADER_SIZE]
        json_len = int(header.decode("ascii"))
        json_bytes = data[HEADER_SIZE : HEADER_SIZE + json_len]

        while len(json_bytes) < json_len:
            more = self.recv_raw()
            json_bytes += more

        return json.loads(json_bytes[:json_len].decode("utf-8"))


def do_key_exchange(
    sock: socket.socket,
    our_static_sk: bytes,
    their_static_pk: bytes,
) -> EncryptedChannel:
    """Perform the ephemeral key exchange after init."""
    server_exchange = read_exact(sock, NONCE_LEN + ENCRYPTED_BLOCK_SIZE)
    server_nonce = server_exchange[:NONCE_LEN]
    server_ct = server_exchange[NONCE_LEN:]

    server_block = crypto_box_open(server_ct, server_nonce, their_static_pk, our_static_sk)

    server_recv_pk = server_block[:KEY_LEN]
    server_send_pk = server_block[KEY_LEN : KEY_LEN * 2]

    our_recv_pk, our_recv_sk = generate_x25519_keypair()
    our_send_pk, our_send_sk = generate_x25519_keypair()

    block = bytearray(BLOCK_SIZE)
    block[:KEY_LEN] = our_recv_pk
    block[KEY_LEN : KEY_LEN * 2] = our_send_pk
    block[KEY_LEN * 2 :] = os.urandom(BLOCK_SIZE - KEY_LEN * 2)

    client_nonce = os.urandom(NONCE_LEN)
    client_ct = crypto_box(bytes(block), client_nonce, their_static_pk, our_static_sk)

    sock.sendall(client_nonce + client_ct)

    return EncryptedChannel(
        sock=sock,
        ours_recv_sk=our_recv_sk,
        ours_send_sk=our_send_sk,
        theirs_recv_pk=server_send_pk,
        theirs_send_pk=server_recv_pk,
    )

File: lua/numscull/crypto.lua
-- numscull/crypto.lua — NaCl Box encryption, key exchange, encrypted channel
-- Port of mockscull/src/numscull/crypto.py using LuaJIT FFI to libsodium

local M = {}
local ffi = require("ffi")

-- Constants from mockscull transport/crypto
M.HEADER_SIZE = 10
M.BLOCK_SIZE = 512
M.MAC_BYTES = 16
M.NONCE_LEN = 24
M.KEY_LEN = 32
M.ENCRYPTED_BLOCK_SIZE = M.BLOCK_SIZE + M.MAC_BYTES -- 528

ffi.cdef([[
  int sodium_init(void);
  void randombytes_buf(void *buf, size_t size);
  int crypto_scalarmult_base(unsigned char *q, const unsigned char *n);
  int crypto_box_easy(unsigned char *c, const unsigned char *m,
                      unsigned long long mlen, const unsigned char *n,
                      const unsigned char *pk, const unsigned char *sk);
  int crypto_box_open_easy(unsigned char *m, const unsigned char *c,
                           unsigned long long clen, const unsigned char *n,
                           const unsigned char *pk, const unsigned char *sk);
]])

local sodium
do
  local names = { "sodium", "libsodium" }
  for _, name in ipairs(names) do
    local ok, err = pcall(function()
      sodium = ffi.load(name)
    end)
    if ok and sodium then break end
    if not ok then M._load_error = tostring(err) end
  end
  if not sodium then
    sodium = nil
  end
end

if not sodium then
  return M
end

-- 0 = first init, 1 = already initialized (idempotent), -1 = failure
local init_ret = sodium.sodium_init()
if init_ret == -1 then
  M._load_error = "sodium_init failed"
  return M
end

-- Pack little-endian u16
local function pack_u16_le(n)
  n = n % 65536
  return string.char(n % 256, math.floor(n / 256))
end

-- Unpack little-endian u16
local function unpack_u16_le(s)
  local b1, b2 = string.byte(s, 1, 2)
  return b1 + b2 * 256
end

-- Pack little-endian u64 (8 bytes) for nonce
local function pack_u64_le(n)
  local s = {}
  for _ = 1, 8 do
    s[#s + 1] = string.char(n % 256)
    n = math.floor(n / 256)
  end
  return table.concat(s)
end

--- Encode counter as 24-byte nonce: LE u64 + 16 zero bytes.
function M.counter_nonce(counter)
  return pack_u64_le(counter) .. string.rep("\0", 16)
end

--- Generate X25519 keypair. Returns (public_key, secret_key). Each 32 bytes.
function M.generate_x25519_keypair()
  local sk = ffi.new("unsigned char[32]")
  sodium.randombytes_buf(sk, 32)
  local pk = ffi.new("unsigned char[32]")
  if sodium.crypto_scalarmult_base(pk, sk) ~= 0 then
    error("crypto_scalarmult_base failed")
  end
  return ffi.string(pk, 32), ffi.string(sk, 32)
end

--- Load keypair from identity file. Returns (public_key, secret_key). Each 32 bytes.
--- config_dir: path string (e.g. "/path/to/config")
--- identity_name: string (e.g. "python-client")
function M.load_keypair(identity_name, config_dir)
  local path = config_dir .. "/identities/" .. identity_name
  local f = io.open(path, "rb")
  if not f then
    error("cannot open identity file: " .. path)
  end
  local raw = f:read("*a")
  f:close()
  if #raw ~= 64 then
    error("expected 64-byte identity file, got " .. #raw)
  end
  return raw:sub(1, 32), raw:sub(33, 64)
end

--- Pack plaintext bytes with 10-byte zero-padded decimal length header.
function M.pack_plaintext_bytes(payload)
  local len_str = string.format("%010d", #payload)
  return len_str .. payload
end

--- EncryptedChannel: encrypted communication using ephemeral X25519 + NaCl Box.
--- Created by do_key_exchange. Use :send(msg) and :recv().
function M.EncryptedChannel(sock, ours_recv_sk, ours_send_sk, theirs_recv_pk, theirs_send_pk)
  local send_nonce = 1
  local recv_nonce = 1

  local function send(message)
    local json_str = vim.json.encode(message)
    local json_bytes = json_str
    local framed = M.pack_plaintext_bytes(json_bytes)
    if #framed > M.BLOCK_SIZE - 2 then
      error("message too large: " .. #framed .. " > " .. (M.BLOCK_SIZE - 2))
    end

    local block = ffi.new("unsigned char[?]", M.BLOCK_SIZE)
    ffi.copy(block, pack_u16_le(#framed), 2)
    ffi.copy(block + 2, framed, #framed)
    -- Random padding for remaining bytes
    local padding_len = M.BLOCK_SIZE - 2 - #framed
    if padding_len > 0 then
      local pad = ffi.new("unsigned char[?]", padding_len)
      sodium.randombytes_buf(pad, padding_len)
      ffi.copy(block + 2 + #framed, pad, padding_len)
    end

    local nonce = M.counter_nonce(send_nonce)
    send_nonce = send_nonce + 1

    local ct = ffi.new("unsigned char[?]", M.BLOCK_SIZE + M.MAC_BYTES)
    local n = ffi.new("unsigned char[24]")
    ffi.copy(n, nonce, 24)
    local their_pk = ffi.new("unsigned char[32]")
    ffi.copy(their_pk, theirs_send_pk, 32)
    local our_sk = ffi.new("unsigned char[32]")
    ffi.copy(our_sk, ours_send_sk, 32)

    if sodium.crypto_box_easy(ct, block, M.BLOCK_SIZE, n, their_pk, our_sk) ~= 0 then
      error("crypto_box_easy failed")
    end

    sock:write(ffi.string(ct, M.ENCRYPTED_BLOCK_SIZE))
  end

  local function recv_raw()
    local ct = sock:read(M.ENCRYPTED_BLOCK_SIZE)
    if not ct or #ct ~= M.ENCRYPTED_BLOCK_SIZE then
      error("socket closed or short read")
    end

    local nonce = M.counter_nonce(recv_nonce)
    recv_nonce = recv_nonce + 1

    local block = ffi.new("unsigned char[?]", M.BLOCK_SIZE)
    local ct_buf = ffi.new("unsigned char[?]", M.ENCRYPTED_BLOCK_SIZE)
    ffi.copy(ct_buf, ct, M.ENCRYPTED_BLOCK_SIZE)
    local n = ffi.new("unsigned char[24]")
    ffi.copy(n, nonce, 24)
    local their_pk = ffi.new("unsigned char[32]")
    ffi.copy(their_pk, theirs_recv_pk, 32)
    local our_sk = ffi.new("unsigned char[32]")
    ffi.copy(our_sk, ours_recv_sk, 32)

    if sodium.crypto_box_open_easy(block, ct_buf, M.ENCRYPTED_BLOCK_SIZE, n, their_pk, our_sk) ~= 0 then
      error("crypto_box_open_easy failed (decryption)")
    end

    local msg_len = unpack_u16_le(ffi.string(block, 2))
    return ffi.string(block + 2, msg_len)
  end

  local function recv()
    local data = recv_raw()
    local header = data:sub(1, M.HEADER_SIZE)
    local json_len = tonumber(header)
    local json_bytes = data:sub(M.HEADER_SIZE + 1, M.HEADER_SIZE + json_len)

    while #json_bytes < json_len do
      local more = recv_raw()
      json_bytes = json_bytes .. more
    end

    return vim.json.decode(json_bytes:sub(1, json_len))
  end

  return {
    send = send,
    recv = recv,
    recv_raw = recv_raw,
  }
end

--- Perform the ephemeral key exchange after init.
--- sock: transport object with read(n) and write(data) methods
--- our_static_sk: 32-byte secret key
--- their_static_pk: 32-byte server public key
--- Returns EncryptedChannel table.
function M.do_key_exchange(sock, our_static_sk, their_static_pk)
  local server_exchange = sock:read(M.NONCE_LEN + M.ENCRYPTED_BLOCK_SIZE)
  if not server_exchange or #server_exchange ~= M.NONCE_LEN + M.ENCRYPTED_BLOCK_SIZE then
    error("short read during key exchange")
  end

  local server_nonce = server_exchange:sub(1, M.NONCE_LEN)
  local server_ct = server_exchange:sub(M.NONCE_LEN + 1)

  local server_block = ffi.new("unsigned char[?]", M.BLOCK_SIZE)
  local ct_buf = ffi.new("unsigned char[?]", M.ENCRYPTED_BLOCK_SIZE)
  ffi.copy(ct_buf, server_ct, M.ENCRYPTED_BLOCK_SIZE)
  local n = ffi.new("unsigned char[24]")
  ffi.copy(n, server_nonce, 24)
  local their_pk = ffi.new("unsigned char[32]")
  ffi.copy(their_pk, their_static_pk, 32)
  local our_sk = ffi.new("unsigned char[32]")
  ffi.copy(our_sk, our_static_sk, 32)

  if sodium.crypto_box_open_easy(server_block, ct_buf, M.ENCRYPTED_BLOCK_SIZE, n, their_pk, our_sk) ~= 0 then
    error("crypto_box_open_easy failed during key exchange")
  end

  local server_recv_pk = ffi.string(server_block, M.KEY_LEN)
  local server_send_pk = ffi.string(server_block + M.KEY_LEN, M.KEY_LEN)

  local our_recv_pk, our_recv_sk = M.generate_x25519_keypair()
  local our_send_pk, our_send_sk = M.generate_x25519_keypair()

  local block = ffi.new("unsigned char[?]", M.BLOCK_SIZE)
  ffi.copy(block, our_recv_pk, M.KEY_LEN)
  ffi.copy(block + M.KEY_LEN, our_send_pk, M.KEY_LEN)
  local pad = ffi.new("unsigned char[?]", M.BLOCK_SIZE - M.KEY_LEN * 2)
  sodium.randombytes_buf(pad, M.BLOCK_SIZE - M.KEY_LEN * 2)
  ffi.copy(block + M.KEY_LEN * 2, pad, M.BLOCK_SIZE - M.KEY_LEN * 2)

  local client_nonce = ffi.new("unsigned char[24]")
  sodium.randombytes_buf(client_nonce, M.NONCE_LEN)

  local client_ct = ffi.new("unsigned char[?]", M.ENCRYPTED_BLOCK_SIZE)
  local their_pk2 = ffi.new("unsigned char[32]")
  ffi.copy(their_pk2, their_static_pk, 32)
  local our_sk2 = ffi.new("unsigned char[32]")
  ffi.copy(our_sk2, our_static_sk, 32)

  if sodium.crypto_box_easy(client_ct, block, M.BLOCK_SIZE, client_nonce, their_pk2, our_sk2) ~= 0 then
    error("crypto_box_easy failed")
  end

  sock:write(ffi.string(client_nonce, M.NONCE_LEN) .. ffi.string(client_ct, M.ENCRYPTED_BLOCK_SIZE))

  return M.EncryptedChannel(sock, our_recv_sk, our_send_sk, server_send_pk, server_recv_pk)
end

--- Write a keypair to disk (identities/<name> and users/<name>.pub for server).
--- Used by tests to generate test identity.
function M.write_keypair(config_dir, identity_name, public_key, secret_key)
  local identities_dir = config_dir .. "/identities"
  local users_dir = config_dir .. "/users"
  vim.fn.mkdir(identities_dir, "p")
  vim.fn.mkdir(users_dir, "p")
  local identity_path = identities_dir .. "/" .. identity_name
  local f = io.open(identity_path, "wb")
  if not f then
    error("cannot write identity file: " .. identity_path)
  end
  f:write(public_key)
  f:write(secret_key)
  f:close()
  local pub_path = users_dir .. "/" .. identity_name .. ".pub"
  local pf = io.open(pub_path, "wb")
  if pf then
    pf:write(public_key)
    pf:close()
  end
end

return M
